components/script/fetch.rs

```diff
@@ -337,11 +337,16 @@ fn fill_headers_with_metadata(r: DomRoot<Response>, m: Metadata, can_gc: CanGc)
     r.set_redirected(m.redirected);
 }
 
+pub(crate) trait CspViolationsProcessor {
+    fn process_csp_violations(&self, violations: Vec<Violation>);
+}
+
 /// Convenience function for synchronously loading a whole resource.
-pub(crate) fn load_whole_resource(
+pub(crate) fn load_whole_resource<P: CspViolationsProcessor>(
```

Comment:
We can use `dyn` instead of a generic type here. That will reduce code size and the performance difference is not important here.